src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/VTableSliceNode.cs

```diff
@@ -135,4 +76,106 @@ public override bool HasConditionalStaticDependencies
             }
         }
     }
+
+    /// <summary>
+    /// Represents a VTable slice for a complete type - a type with all virtual method slots generated,
+    /// irrespective of whether they are used.
+    /// </summary>
+    internal sealed class CompleteVTableSliceNode : VTableSliceNode
+    {
+        private MethodDesc[] _slots;
+
+        public CompleteVTableSliceNode(TypeDesc type)
+            : base(type)
+        {
+            var slots = new ArrayBuilder<MethodDesc>();
+
+            MetadataType mdType = _type.GetClosestMetadataType();
+            foreach (var method in mdType.GetMethods())
+            {
+                if (!method.IsVirtual)
+                    continue;
+
+                slots.Add(method);
+            }
+
+            _slots = slots.ToArray();
+        }
+
+        public override IReadOnlyList<MethodDesc> Slots
+        {
+            get
+            {
+                return _slots;
+            }
+        }
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory factory)
+        {
+            List<DependencyListEntry> dependencies = new List<DependencyListEntry>(_slots.Length + 1);
+            if (_type.HasBaseType)
+            {
+                dependencies.Add(new DependencyListEntry(factory.VTable(_type.BaseType), "Base type VTable"));
+            }
+
+            foreach (MethodDesc method in _slots)
+            {
+                dependencies.Add(new DependencyListEntry(factory.VirtualMethodUse(method), "Full vtable dependency"));
+            }
+
+            return dependencies;
+        }
+    }
+
+    /// <summary>
+    /// Represents a VTable slice where slots are built on demand. Only the slots that are actually used
+    /// will be generated.
+    /// </summary>
+    internal sealed class DemandBuiltVTableSliceNode : VTableSliceNode
```

Comment:
DemandBuilt would be better phrased as OnDemandBuild, or perhaps EagerlyBuilt vs LazilyBuilt?